### PR TITLE
oslo.messaging driver

### PR DESCRIPTION
**Pull request by iurygregory**
This PR adds a new driver for oslo.messaging notifier to capture the messages and write to a file.
The driver is installed under the oslo.messaging namespace.

To cofigure the driver is necessary to edit the `oslo_messaging_notifications` section in `ironic.conf` 
```
[oslo_messaging_notifications]
driver = prometheus_exporter
transport_url = fake://
file_dir=/tmp/ironic_prometheus_exporter
file_name=myfile.txt
```

### CODE REVIEW

File: README.md
Comment (dtantsur) [inline README.md]: nit: this should be under the top heading, not on the same level

File: README.md
Comment (dtantsur) [inline README.md:34]: Pardon me if it's a stupid question, but can we merge `file_dir` and `file_name` into `transport_url`, like
```
transport_url=file:///tmp/ironic_prometheus_exporter/myfile.txt
```
?

File: ironic_prometheus_exporter/messaging.py
Comment (dtantsur) [inline ironic_prometheus_exporter/messaging.py]: nit: redundant empty line

File: ironic_prometheus_exporter/messaging.py
Comment (dtantsur) [inline ironic_prometheus_exporter/messaging.py]: nit: use `os.path.join`

File: requirements.txt
Comment (dtantsur) [inline requirements.txt]: You don't seem to import this anywhere.

File: ironic_prometheus_exporter/messaging.py
Comment (iurygregory) [reply ironic_prometheus_exporter/messaging.py]: ack

File: requirements.txt
Comment (iurygregory) [reply requirements.txt]: Oh, I forgot to remove.

File: ironic_prometheus_exporter/messaging.py
Comment (iurygregory) [reply ironic_prometheus_exporter/messaging.py]: ack

File: README.md
Comment (iurygregory) [reply README.md]: ack

File: README.md
Comment (iurygregory) [reply README.md:34]: When I talked with Julia in the begin our idea was use something like this in the `transport_url`, but while testing I ended up with 
```
oslo_messaging.transport.DriverLoadFailure: Failed to load transport driver "file": No \'oslo.messaging.drivers\' driver found, looking for \'file\''
```
To use transport_url will be necessary create a Transport Driver

Comment (dtantsur) [reply README.md:34]: Ugh. Okay, let's leave it as it is. Why do you split `file_dir` and `file_name` though?

Comment (iurygregory) [reply README.md:34]: The reason I've split was just to let the driver verify if the directory exists and if not it will create

Comment (cybertron) [reply README.md:34]: You could do that with os.path.dirname on the full path.

File: ironic_prometheus_exporter/messaging.py
Comment (cybertron) [inline ironic_prometheus_exporter/messaging.py]: Nit: This won't function as a docstring. It should use the """ notation.

"""Publish notifications into a File to be used by Prometheus"""

Comment (iurygregory) [reply ironic_prometheus_exporter/messaging.py]: ack

File: ironic_prometheus_exporter/messaging.py
Comment (dhellmann) [inline ironic_prometheus_exporter/messaging.py]: It doesn't look like this is going to produce the right file format. Isn't the point of this plugin to generate the prometheus formatted data to be served by something else?

Comment (iurygregory) [reply ironic_prometheus_exporter/messaging.py]: Hi Doug, if I do remember we want the driver as simple as possible, we will format the data in the file after is written

Comment (dhellmann) [reply ironic_prometheus_exporter/messaging.py]: That's  not what I remember from the conversation at all. The idea was to have the notifier emit the data in a format that is ready to be served, and then provide a very simple web service to allow prometheus to ask for the data.
If you're going to store the message in some format for later processing, you need to use JSON or some other encoding system in order to have parseable data.

Comment (dtantsur) [reply ironic_prometheus_exporter/messaging.py]: My recollections mathes Doug's. I thought this patch was just the first step and the next patches would introduce the correct format.

Comment (juliakreger) [reply ironic_prometheus_exporter/messaging.py]: So Doug and Dmitry's recollections don't match Iury's and my recollections.

The orignally discussed point, and as written in the stories this work is based upon is to a) collect the data, and once we have all of it, whatever is exposing that data to prometheus's web browser (because prometheus may not visit that often as the metrics are updated), we then collect all of the files and transform it to a prometheus friendly format before handing back over the webserver socket. I think I view the filtering as a heavy-ish load but still a very simple thing because it doesn't need to be super featureful, it just needs to iterate through data and return what it has. Where as this is going to be writing files per node to some folder with no solid consistency between drivers so for each driver we're going to have to add transformations as we go.

Comment (iurygregory) [reply ironic_prometheus_exporter/messaging.py]: So I should keep the driver simple just writing the messages into a json file? What you everyone think about it? I have no problem to change to do the parser for the messages too, but this would require some samples of the message format etc.

Comment (juliakreger) [reply ironic_prometheus_exporter/messaging.py]: I think it would be best, but we could just do it all in the plugin. :\

To get the sample data, I think we should use this plugin to dump the data out on one of the dev machines we have. To do so, we should loop in @derekhiggins. Worst comes to worst, I can power on my supermicro machine in my house and get it back online, although the message data will differ being a different vendor than our initial target.

Comment (iurygregory) [reply ironic_prometheus_exporter/messaging.py]: So maybe we can go with the following workflow:
1- update the plugin here to write to a json file
2- Try to use the plugin in a real env so we can collect the data (Only one node is fine for this, but I would say that two would make thinks easier since we would be able to map correctly for N nodes)
3- Update the plugin to parse the data 
What everyone think about this?

Comment (dhellmann) [reply ironic_prometheus_exporter/messaging.py]: That plan works for me.

Comment (juliakreger) [reply ironic_prometheus_exporter/messaging.py]: I think this path brings us forward faster, and from a transformation standpoint, we could likely take a list of the objects we want and just being transforming them as we go or dropping them on the floor.

Comment (iurygregory) [reply ironic_prometheus_exporter/messaging.py]: Thanks everyone, changes have been made =)

Comment (derekhiggins) [reply ironic_prometheus_exporter/messaging.py]: after talking to @iurygregory, he is going to use the plugin to gather real metrics once the initial version of the plugin merged.

File: ironic_prometheus_exporter/messaging.py
Comment (etingof) [inline ironic_prometheus_exporter/messaging.py:22]: nit: there is also `os.path.splitext()`

File: ironic_prometheus_exporter/messaging.py
Comment (etingof) [inline ironic_prometheus_exporter/messaging.py:28]: Would it make sense to loop over (potentially) failing write respecting the `retry` parameter?

Comment (iurygregory) [reply ironic_prometheus_exporter/messaging.py:28]: Hi Ilya, the notify method is an abstract method defined in the Driver class in oslo.messaging [1]
so we keep the same parameters 

[1] https://github.com/openstack/oslo.messaging/blob/master/oslo_messaging/notify/notifier.py#L125

Comment (etingof) [reply ironic_prometheus_exporter/messaging.py:28]: o/
Right, we are implementing this method, it has the `retry` parameter:

```
        :param retry: connection retries configuration (used by the messaging
                      driver):
                      None or -1 means to retry forever.
                      0 means no retry is attempted.
                      N means attempt at most N retries.
```

It seems we currently ignore `retry`, so I wonder if we should respect it in one way or the other...?

Comment (iurygregory) [reply ironic_prometheus_exporter/messaging.py:28]: This is the initial version for the driver, we just want the messages written to the json file, I think it's  not necessary for now. For example the log driver doesn't use

https://github.com/openstack/oslo.messaging/blob/648ad56c8652d83d3b1cae745488bb8e0450febd/oslo_messaging/notify/_impl_log.py

Comment (dtantsur) [reply ironic_prometheus_exporter/messaging.py:28]: I don't think we expect writing to local files to fail in a retryable way :)

File: ironic_prometheus_exporter/messaging.py
Comment (dtantsur) [inline ironic_prometheus_exporter/messaging.py:23]: I'd drop this check, why not call the file the way users want?

File: ironic_prometheus_exporter/tests/test_driver.py
Comment (dtantsur) [inline ironic_prometheus_exporter/tests/test_driver.py:10]: Let's add a real unit test on writing something

**Review by dtantsur — APPROVED** [review]

File: ironic_prometheus_exporter/messaging.py
Comment (etingof) [reply ironic_prometheus_exporter/messaging.py:28]: Well, "No space left on device" seems theoretically possible and potentially recoverable..

I am just thinking how could we behave in a way the abstract interface expects us to...

Comment (dtantsur) [reply ironic_prometheus_exporter/messaging.py:28]: Yes, but retrying won't help until somebody cleans up the disk :)

File: ironic_prometheus_exporter/messaging.py
Comment (iurygregory) [reply ironic_prometheus_exporter/messaging.py:23]: Yeah make sense, since we will process the message and write in the prometheus format

File: ironic_prometheus_exporter/tests/test_driver.py
Comment (iurygregory) [reply ironic_prometheus_exporter/tests/test_driver.py:10]: Ack

File: ironic_prometheus_exporter/messaging.py
Comment (etingof) [reply ironic_prometheus_exporter/messaging.py:28]: Ironic lands are inhabited by daemons. One might clean things up, eventually. Also, the work flow can consume file system space temporarily.

Just thinking aloud, that's nit indeed.

File: ironic_prometheus_exporter/messaging.py
Comment (juliakreger) [reply ironic_prometheus_exporter/messaging.py:23]: Once we have the data, @iurygregory can confirm that he is getting a node name or uuid across and we can change the path so we just overwrite data  as we go for the existing file. I think we will get it in one huge latest chunk per node so we should always be safe to replace the prior node's data.

File: README.md
Comment (juliakreger) [inline README.md:35]: Ultimately a folder would be best because prometheus is only going to come along every five minutes or so.

**Review by juliakreger — APPROVED** [review]
To get started, LGTM.

File: README.md
Comment (iurygregory) [reply README.md:35]: So the file name/format doesn't matter?

Comment (juliakreger) [reply README.md:35]: nope, we should just be dumping content into a folder by some unique identifier in the message, per message and replacing any prior file.

File: ironic_prometheus_exporter/messaging.py
Comment (iurygregory) [reply ironic_prometheus_exporter/messaging.py:23]: Also do we want a default path ? When I was talking with Mike for the test plan he asked if would be really necessary to setup configuration for the driver or if we have default values.

Comment (juliakreger) [reply ironic_prometheus_exporter/messaging.py:23]: I'm sure we're going to need to allow it to be configurable, but I don't know about the default path. I think we can change that if it makes sense to do so later... Maybe we even dump that into the shared folder, but that means ironic configuration that is static once we begin integrating... since we will also need to turn on metrics collection.

**Pull request merged**
